core/src/main/java/org/apache/druid/segment/column/TypeStrategy.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Most implementations of this mechanism, support reading and writing ONLY non-null values. The exception to this is
+ * {@link NullableTypeStrategy}, which might be acceptable to use if you need to read and write nullable values, AND,
+ * you have enough memory to burn a full byte for every value you want to store. It will store values with a leading
+ * byte containing either {@link NullHandling#IS_NULL_BYTE} or {@link NullHandling#IS_NOT_NULL_BYTE} as appropriate.
+ * If you have a lot of values to write and a lot of nulls, consider alternative approaches to tracking your nulls.
+ *
+ * This mechanism allows using the natural {@link ByteBuffer#position()} and modify the underlying position as they
+ * operate, and also random access reads are specific offets, which do not modify the underlying position. If a method
+ * accepts an offset parameter, it does not modify the position, if not, it does.
+ *
+ * The only methods implementors are required to provide are {@link #read(ByteBuffer)},
+ * {@link #write(ByteBuffer, Object, int)} and {@link #estimateSizeBytes(Object)}, default implementations are provided
+ * to set and reset buffer positions as appropriate for the offset based methods, but may be overridden if a more
+ * optimized implementation is needed.
+ */
+public interface TypeStrategy<T> extends Comparator<T>
+{
+  /**
+   * Estimate the size in bytes that writing this value to memory would require. This method is not required to be
+   * exactly correct, but many implementations might be. Implementations should err on the side of over-estimating if
+   * exact sizing is not efficient
+   */
+  int estimateSizeBytes(T value);
+
+
+  /**
+   * Read a non-null value from the {@link ByteBuffer} at the current {@link ByteBuffer#position()}. This will move
+   * the underlying position by the size of the value read.
+   *
+   * The contract of this method is that any value returned from this method MUST be completely detached from the
+   * underlying {@link ByteBuffer}, since it might outlive the memory location being allocated to hold the object.
+   * In other words, if an object is memory mapped, it must be copied on heap, or relocated to another memory location
+   * that is owned by the caller with {@link #write}.
+   */
+  T read(ByteBuffer buffer);
+
+  /**
+   * Write a non-null value to the {@link ByteBuffer} at position {@link ByteBuffer#position()}. This will move the
+   * underlying position by the size of the value written, and returns the number of bytes written.
+   *
+   * If writing the value would take more than 'maxSizeBytes' (or the buffer limit, whichever is smaller), this method
+   * will return a negative value indicating the number of additional bytes that would be required to fully write the
+   * value. Partial results may be written to the buffer when in this state, and the position may be left at whatever
+   * point the implementation ran out of space while writing the value.
```

Comment:
Does this mean the caller is expected to save the position? Or is the caller safe to assume it can roll back by `-retval`?

i.e., is this code safe?

```
int written = write(buf, value, maxBytes);
if (written < 0) {
  buf.position(buf.position() + written); // roll back buf, undoing the write
}
```

or do callers need to do this?

```
int oldPosition = buf.position();
int written = write(buf, value, maxBytes);
if (written < 0) {
  buf.position(oldPosition); // roll back buf, undoing the write
}
```

Reply:
the latter, caller should save the position before writing to handle an incomplete write, added javadocs to clarify